### PR TITLE
[RLlib] DreamerV3 on PyTorch.

### PR DESCRIPTION
**Pull request by sven1977**
DreamerV3 on PyTorch.

The RLlib team finally managed to fully transfer DreamerV3 to PyTorch, including support for multi-GPU training.

Training (and sampling) throughput nicely scale linearly now with the number of GPUs/Learners used:

image

Learning speed should also scale, albeit most likely sub-linearly. Learning rate adjustments are crucial to get right. Below is a 4GPU run with learning rates 4x and a 2GPU run with learning rates 1.4x (x2**0.5).
image

For 8 GPUs or more linear scaling for Pong becomes more difficult (but this may not apply to harder envs), but more complex lr-schedules may help.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/BUILD [hunk 1]
```diff
@@ -1482,6 +1482,13 @@ py_test(
     srcs = ["utils/exploration/tests/test_random_encoder.py"]
 )
 
+py_test(
```
Comment (sven1977) [inline rllib/BUILD:1485]: added `two_hot` tests (for torch, too, of course).

File: rllib/algorithms/dreamerv3/dreamerv3.py [hunk 2]
```diff
@@ -28,24 +26,29 @@
 from ray.rllib.core import DEFAULT_MODULE_ID
 from ray.rllib.core.columns import Columns
 from ray.rllib.core.rl_module.rl_module import SingleAgentRLModuleSpec
+from ray.rllib.execution.rollout_ops import synchronous_parallel_sample
```
Comment (sven1977) [inline rllib/algorithms/dreamerv3/dreamerv3.py:29]: This is the most important file to review, the rest is mere "translation" from the existing tf version to torch.

File: rllib/algorithms/dreamerv3/utils/summaries.py [hunk 3]
```diff
@@ -115,7 +116,7 @@ def report_dreamed_trajectory(
 
 def report_predicted_vs_sampled_obs(
     *,
-    results,
+    metrics,
```
Comment (sven1977) [inline rllib/algorithms/dreamerv3/utils/summaries.py:119]: We pass in the main Algo's metrics object now and just log into that (instead of returning altered dicts).

File: rllib/core/learner/learner.py [hunk 4]
```diff
@@ -1045,36 +1041,13 @@ def additional_update_for_module(
...
-                self.metrics.log_value(
-                    key=(module_id, stats_name), value=new_lr, window=1
-                )
+        pass
 
     def update_from_batch(
         self,
         batch: MultiAgentBatch,
         *,
+        timesteps=None,  # TODO
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1050]: this is still a TODO for this PR:
* We should pass in the current global timesteps (env, agents and modules) into all `update` calls
* instead, we skip the `additional_update` call and replace it with a user-overridable before_update and after_update sub-method, called automatically by `Learner.update_from_batch|episodes`.
* In the future, we might simplify by deprecating `update_from_batch` and fully replace it with `update`, which will take only episodes. ??

File: rllib/core/learner/learner.py [hunk 5]
```diff
@@ -1388,22 +1356,36 @@ def _update_from_batch_or_episodes(
...
-            {
-                (mid, f"{full_name[len(mid) + 1 :]}_lr"): convert_to_numpy(
-                    self._get_optimizer_lr(self._named_optimizers[full_name])
-                )
-                for mid, full_names in self._module_optimizers.items()
-                for full_name in full_names
-            },
-            window=1,
-        )
+        self._after_update(timesteps)
```
Comment (sven1977) [inline rllib/core/learner/learner.py:1359]: maybe make this public

File: rllib/core/models/torch/utils.py [hunk 6]
```diff
@@ -45,28 +45,33 @@ def __init__(self, width, height, stride_w, stride_h):
         self.stride_w = stride_w
         self.stride_h = stride_h
 
-        self.zeros = torch.zeros(
-            size=(
-                self.width * self.stride_w - (self.stride_w - 1),
-                self.height * self.stride_h - (self.stride_h - 1),
+        self.register_buffer(
```
Comment (sven1977) [inline rllib/core/models/torch/utils.py:48]: make this GPU-capable.

File: rllib/core/learner/learner.py (hunk 4 above)
Comment (simonsays1980) [reply rllib/core/learner/learner.py:1050]: As we fully focus on episodes this makes sense in my opinion.

File: rllib/core/learner/learner.py [hunk 7]
```diff
@@ -1388,22 +1356,36 @@ def _update_from_batch_or_episodes(
...
 
         # Reduce results across all minibatch update steps.
         return self.metrics.reduce()
 
+    # TODO (sven): Document and design this new approach in order to replace
+    #  `additional_update()`.
+    def _after_update(self, timesteps):
```
Comment (simonsays1980) [inline rllib/core/learner/learner.py:1366]: I thought this function needs to be public, but we are using also `_forward_train`, etc. for overriding in the `RLModule` .... so this should be fine.

File: rllib/tuned_examples/dreamerv3/atari_200M.py [hunk 8]
```diff
@@ -50,10 +50,21 @@
...
+    )
+    .env_runners(
+        num_env_runners=(args.num_env_runners or 0),
+        # If we use >1 GPU and increase the batch size accordingly, we should also
+        # increase the number of envs per worker.
+        num_envs_per_env_runner=8 * (args.num_gpus or 1),
+        remote_worker_envs=True,
+    )
+    .learners(
+        num_learners=0 if args.num_gpus == 1 else args.num_gpus,
```
Comment (simonsays1980) [inline rllib/tuned_examples/dreamerv3/atari_200M.py:63]: Btw, isn't it possible to use here fractional GPUs as well? Even though the model is large I guess a V100 could be shared between 2 learners?

File: rllib/utils/torch_utils.py [hunk 9]
```diff
@@ -684,6 +684,129 @@ def softmax_cross_entropy_with_logits(
     return torch.sum(-labels * nn.functional.log_softmax(logits, -1), -1)
 
 
+@PublicAPI
+def symlog(x: "torch.Tensor") -> "torch.Tensor":
```
Comment (simonsays1980) [inline rllib/utils/torch_utils.py:688]: Can't we compile such torch functions?

File: rllib/utils/torch_utils.py [hunk 10]
```diff
@@ -684,6 +684,129 @@ def softmax_cross_entropy_with_logits(
...
+    # at the given indices.
+    output = torch.zeros(value.shape[0], num_buckets, device=device)
+    # Set our two-hot values at computed indices.
+    output[indices[:, 0], indices[:, 1]] = updates
+    return output
+
+
 def _dynamo_is_available():
     # This only works if torch._dynamo is available
     try:
```
Comment (simonsays1980) [inline rllib/utils/torch_utils.py:812]: Btw is this now available on CI?

**Review by simonsays1980 — APPROVED** [review]
LGTM. Awesome PR. So great to have it now in PyTorch!

File: rllib/algorithms/dreamerv3/torch/models/dreamer_model.py [hunk 11]
```diff
@@ -0,0 +1,490 @@
...
+            is_first=is_first,
+        )
+
+    def get_initial_state(self):
+        """Returns the (current) initial state of the dreamer model (a, h-, z-states).
+
+        An initial state is generated using the previous action, the tanh of the
+        (learned) h-state variable and the dynamics predictor (or "prior net") to
+        compute z^0 from h0. In this last step, it is important that we do NOT sample
+        the z^-state (as we would usually do during dreaming), but rather take the mode
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/torch/models/dreamer_model.py:158]: Z-hat (z^) is the "prior" z-state, meaning a z-state that gets computed using only(!) the h-state as input.
On the other hand, a z-state (without the hat) gets computed using the "posterior" net, which has access to the h-state AND the encoder's output.

File: rllib/algorithms/dreamerv3/torch/models/components/representation_layer.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        logits = torch.log(probs)
+
+        # Create the distribution object using the unimix'd logits.
+        # Note that in the tf-version, we use an tfp.distribution.Independent
+        # wrapper, which is unnecessary, b/c we only sample from our distribution.
+        # We do NOT need to compute logps (for the combined 32 categoricals).
+        distribution = torch.distributions.OneHotCategorical(logits=logits)
+
+        # Draw a one-hot sample (B, num_categoricals, num_classes).
+        sample = distribution.sample()
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/torch/models/components/representation_layer.py:122]: I guess, we could do `rsample`, but I felt it was much safer to do the straight-through sampling explicitly here in the code.

File: rllib/algorithms/dreamerv3/torch/dreamerv3_torch_learner.py [hunk 13]
```diff
@@ -0,0 +1,913 @@
...
+torch, nn = try_import_torch()
+
+
+class DreamerV3TorchLearner(DreamerV3Learner, TorchLearner):
+    """Implements DreamerV3 losses and gradient-based update logic in TensorFlow.
+
+    The critic EMA-copy update step can be found in the `DreamerV3Learner` base class,
+    as it is framework independent.
+
+    We define 3 local TensorFlow optimizers for the sub components "world_model",
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/torch/dreamerv3_torch_learner.py:34]: great catch. Fixed.

File: rllib/algorithms/dreamerv3/tests/test_dreamerv3.py [hunk 14]
```diff
@@ -209,32 +242,231 @@ def test_dreamerv3_dreamer_model_sizes(self):
...
+                print(f"Testing model_size={model_size} on env-type: {env_name} ..")
+                config.environment(
+                    observation_space=obs_space,
+                    action_space=gym.spaces.Discrete(num_actions),
+                )
+
+                modules = []
+                tf_weights = []
+
+                for fw in framework_iterator(config, frameworks=("tf2", "torch")):
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/tests/test_dreamerv3.py:328]: We will, but first, we need to make the torch version learn Atari :) 
I'll remove these tests once this is all confirmed and then we can remove the tf2 stuff as well.

File: rllib/algorithms/dreamerv3/dreamerv3.py [hunk 15]
```diff
@@ -631,71 +671,50 @@ def training_step(self) -> ResultDict:
...
         # Add train results and the actual training ratio to stats. The latter should
         # be close to the configured `training_ratio`.
-        results.update(train_results)
-        results[ALL_MODULES]["actual_training_ratio"] = self.training_ratio
+        # results.update(train_results)
+        self.metrics.log_value("actual_training_ratio", self.training_ratio, window=1)
 
         # Return all results.
-        return results
+        return self.metrics.reduce()
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3.py]: In the future, we should reduce later (users should NOT call reduce() anymore at all, anywhere). But right now, we still require this here. Should be a separate PR, imo.

File: rllib/algorithms/dreamerv3/dreamerv3.py [hunk 16]
```diff
@@ -535,45 +544,74 @@ def training_step(self) -> ResultDict:
...
-                )
+                env_steps_last_regular_sample = sum(len(eps) for eps in episodes)
                 total_sampled = env_steps_last_regular_sample
 
                 # If we have never sampled before (just started the algo and not
                 # recovered from a checkpoint), sample B random actions first.
-                if self._counters[NUM_AGENT_STEPS_SAMPLED] == 0:
-                    d_, o_ = env_runner.sample(
-                        num_timesteps=(
+                if self.metrics.peek(NUM_ENV_STEPS_SAMPLED_LIFETIME, default=0) == 0:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3.py]: Ah, great point. I think you are right. ...

File: rllib/algorithms/dreamerv3/dreamerv3.py [hunk 17]
```diff
@@ -506,8 +517,6 @@ def setup(self, config: AlgorithmConfig):
 
     @override(Algorithm)
     def training_step(self) -> ResultDict:
-        results = {}
-
         env_runner = self.workers.local_worker()
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3.py]: done

File: rllib/algorithms/dreamerv3/dreamerv3.py [hunk 18]
```diff
@@ -535,45 +544,74 @@ def training_step(self) -> ResultDict:
                 or not have_sampled
             ):
                 # Sample using the env runner's module.
-                done_episodes, ongoing_episodes = env_runner.sample()
+                episodes, env_runner_results = synchronous_parallel_sample(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3.py]: see further below: This property always computes the actual current training ratio using the actual counts.
```
@property
def training_ratio(self) -> float:

```

File: rllib/algorithms/dreamerv3/dreamerv3.py [hunk 19]
```diff
@@ -421,7 +420,13 @@ def validate(self) -> None:
 
     @override(AlgorithmConfig)
     def get_default_learner_class(self):
-        if self.framework_str == "tf2":
+        if self.framework_str == "torch":
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/dreamerv3/dreamerv3.py:424]: I wish! Torch needs to learn Atari first. We are close, but it's not learning yet. Must find the "final bug" before we can merge this ...

**Comment by Viibrant** [comment]
Hi all, what is the current status of this PR?
